Add wrapper for NFD encoding workaround

PVince81 opened the pull request:
Fixes https://github.com/owncloud/core/issues/21365

Adds storage wrapper "EncodingWrapper" that first tries to find which one of NFC or NFD file exists, and then uses that path to perform operations.
#### TODO
- [x] TODO: introduce wrapper
- [x] TODO: make wrapper opt-in using mount option
- [x] TODO: unit tests
#### Bugs

Currently works ok-ish with SFTP, need to be tested with all file operations
- [x] BUG: cache entries randomly appear/disappear, possibly some code paths bypassing the wrapper
- [x] BUG: files:scan sometimes makes it disappear from cache
- [x] BUG: two NFD entries in same folder seem to bug
#### Tests
- [x] TEST: download
- [x] TEST: upload multiple files into NFD folder, refresh folder to see its contents
- [x] TEST: upload + overwrite file into NFD folder
- [x] TEST: upload + overwrite file with NFD name (the uploaded file name is NFC)
- [x] TEST: rename
- [x] TEST: move to storage subdir
- [x] TEST: cross-storage move (move out of the ext storage)
- [x] TEST: copy
- [x] TEST: propfind
- [x] TEST: occ files:scan
- [x] TEST: watcher (touch the files remotely then PROPFIND the folder)
- [x] TEST: permanent delete (no trashbin)
- [x] TEST: delete to trash
- [x] TEST: restore from trash
- [x] TEST: versions creation
- [x] TEST: version download
- [x] TEST: version restore
- [x] TEST: encryption
- [x] TEST: work inside folder that has a NFD name
#### External storages
- [x] SFTP
- [x] SMB
- [ ] ...

@icewind1991 @DeepDiver1975 @nickvergessen FYI

Commit:
Add wrapper for NFD encoding workaround

Commit:
Add encoding wrapper as opt-in mount option

The encoding wrapper is now only applied when the mount option is set,
disabled by default.

PVince81 commented:
More progress. More file operations work now and the option is opt-in in the UI now.

PVince81 commented:
- [ ] BUG: delete a file within a folder containing a "bad ümlaut", the file disappears after deletion (possibly rogue rescan)

PVince81 commented:
Arghh, another folder listing diff that needs NFD support: https://github.com/owncloud/core/blob/v9.0.1/lib/private/files/cache/scanner.php#L397

Commit:
Scanner must normalize new children names for cache diff

Since new children from the storage might contain NFD entries, these
must be normalized to NFC to be properly diff'ed with the cache
contents which is always NFC.

This fixes an issue where NFD entries would disappear from the cache
after rescannng for children.

PVince81 commented:
- [ ] TEST: work in folder name like "/nfd_name/nfc_name/nfd_name"

Will not work with current logic that transforms the whole path to NFC/NFD). Need an algo that goes through all path sections and calls `file_exists` there. :disappointed:

PVince81 commented:
- [ ] BUG: cannot upload into a folder with NFD name

Reason is the same as above. Need to change `findPathToUse` to do a `file_exists` on every section.

PVince81 commented:
Added some unit tests, might not be enough to cover all.

dragotin commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,510 @@
+	 * @return string original or converted path
+	 */
+	private function findPathToUse($path) {
+		if ($path !== '' && !$this->isAscii($path)) {
```
Nitpick: the whole function would save indentation if you said
```
if  (path == '' || $this->isAscii($path) ) {
   return $path;
}
```
Wouldn't that be nicer?

dragotin replied:
And: The isAscii funciton initializes a regexp engine which is always not cheap.

Without having investigated details but don't you think a simple for loop over the string checking for non ascii chars is runtime-cheaper?

PVince81 replied:
We got plenty of indentation in stock so no need to save them :wink: 

Yeah, good point. I'll change it next time I touch this good, which is likely to be soon. The algo isn't adequate because every path sections needs its own NFC/NFD trial...

PVince81 commented:
- [ ] TODO: optimize `isAscii`

dragotin commented:
On a more general note, and sorry if my comments miss the facts, I haven't investigated too deeply: The `pathToUse()` function which does the costly conversion is called in most of the low level functions. That looks like it can happen multiple if code like 
```
if( file_exists($path) ) {
   remove_file($path); 
}
```
is done - that would results in calling the conversion function twice maybe. That can eat performance.

The big gun would of course be a class *Path* that has the knowledge about if the encoding was checked already.

PVince81 commented:
@dragotin actually the expensive lookup for each path is only done once, the result is saved in the `$namesCache` array which should prevent redoing the lookup several times.

PVince81 commented:
@dragotin I fixed isAscii to be simpler and saved some indenting.

Also: changed the path finding algo by converting each path section one by one to find either the NFD or NFC form. This is required in case of people having crazy paths like "nfd/nfc/nfd".

- [ ] TODO: some unit tests fail

Also need to redo a bit of manual testing as per the steps in the OP.

PVince81 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,543 @@
+	 *
+	 * @param string $fullPath path to check
+	 *
+	 * @return string original or converted path, or null if none of the forms was found
```
missed fixing this comment

PVince81 commented:
Rebased.
Fixed the unit tests and also fixed many issues with the encoding wrapper itself.

Next up: redo the manual testing to tick all the TEST checkboxes above.

PVince81 commented:
Please review @owncloud/filesystem @nickvergessen @schiesbn

icewind1991 commented on apps/files_external/js/settings.js:
```diff
@@ -33,8 +33,12 @@ var MOUNT_OPTIONS_DROPDOWN_TEMPLATE =
 	'	</div>' +
+	'	<div class="optionRow">' +
+	'		<input id="mountOptionsEncoding" name="encoding_compatibility" type="checkbox" value="true"/>' +
+	'		<label for="mountOptionsEncoding">{{t "files_external" "Enable encoding compatibility (decreases performance)"}}</label>' +
```
"Enable NFD encoding compatibility" or "Enable Mac unicore encoding compatibility" ?

PVince81 replied:
Yeah, I hesitated about that but you're right. In the end it's only about NFD/Mac.

nickvergessen replied:
btw what is this `{{t` magic, does our language extraction tool detect that?

PVince81 replied:
it's a template function. Unfortunately our language extraction doesn't detect it, so there's an ugly workaround here: https://github.com/owncloud/core/pull/24349/files/e6f31f2107ed86e88527bb4ac2f60936c0e57e3a#diff-55300aafc99213431ff789e8b0fb70a7R13

I hope one day we can make it work... https://github.com/owncloud/core/issues/15106

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,568 @@
+	 *
+	 * @return bool true if the string is all ASCII, false otherwise
+	 */
+	private function isAscii($str) {
```
```php
preg_match('/[\x00-\x7F]+/', $str);
```

seems to be faster

https://3v4l.org/PKHN9/perf#tabs vs https://3v4l.org/llN2N/perf#tabs

PVince81 replied:
Okay, I had it in the first place and @dragotin convinced me to replace it with a loop.
If PHP's regexp is faster then nice!

PVince81 replied:
done with a slightly different regexp

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,568 @@
+		$path = '';
+		foreach ($sections as $section) {
+			$path .= $section;
+			$convertedPath = $this->findPathToUseLastSection($path);
```
might as well pass `$path` and `$section` separatly to save having to split it again in the called function

PVince81 replied:
done

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,568 @@
+		// note: no conversion here, method should not be called with non-NFC names!
+		$result = $this->storage->mkdir($path);
+		if ($result) {
+			unset($this->namesCache[$path]);
```
why unset this here, setting `$this->namesCache[$path] = $path` seems to make more sense

PVince81 replied:
fixed

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,568 @@
+	public function file_put_contents($path, $data) {
+		$result = $this->storage->file_put_contents($this->findPathToUse($path), $data);
+		if ($result) {
+			unset($this->namesCache[$path]);
```
why the unset?

PVince81 replied:
if the file did not exist before but the name was already cached... hmmm... you're right, it's useless

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,568 @@
+			return $this->copy($this->findPathToUse($sourceInternalPath), $targetInternalPath);
+		}
+
+		$result = $this->storage->copyFromStorage($sourceStorage, $this->findPathToUse($sourceInternalPath), $targetInternalPath);
```
`$targetInternalPath` should be de-normalized, not `$sourceInternalPath` since the target path is the one on this storage

PVince81 replied:
ok seems I need to add moar unit tests

Commit:
Fixes for encoding wrapper

Improved label
Fixed rename/copy/moveFromStorage/copyFromStorage and added tests
Improved findPathToUse algo

PVince81 commented:
I fixed the code based on comments, please have another look @icewind1991 @nickvergessen

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,546 @@
+		$result = $this->storage->rename($this->findPathToUse($path1), $this->findPathToUse($path2));
+		if ($result) {
+			unset($this->namesCache[$path1]);
+			unset($this->namesCache[$path2]);
```
more unneeded unsets (and the 2 below)

PVince81 replied:
done

icewind1991 commented on lib/private/Files/Storage/Wrapper/Encoding.php:
```diff
@@ -0,0 +1,546 @@
+	public function touch($path, $mtime = null) {
+		$result = $this->storage->touch($this->findPathToUse($path), $mtime);
+		if ($result) {
+			unset($this->namesCache[$path]);
```
same here

PVince81 replied:
done

Commit:
Remove unneeded unsets in encoding wrapper

PVince81 commented:
Went through the test list and all went fine with SMB.
Many of the tests were done where file operations are done inside a NFD folder called "bad-ümlaut", and the second part the operations are done on a NFD file "bad-ümlaut.txt" directly.

PVince81 commented:
Did a quick test with SFTP and it seems to work fine, too :smile:

PVince81 commented:
Rebased again. I hope this PR is acceptable now :smile: 
@icewind1991 @nickvergessen

PVince81 commented:
Test files are here: https://github.com/owncloud/core/issues/21365#issuecomment-176618528

rullzer commented:
:+1:

icewind1991 commented:
:+1:
